### PR TITLE
new function filterHomographyDecompSolutionsByPointNormals

### PR DESCRIPTION
**Pull request by Vaesper**
### This pullrequest changes

This pull request adds a function that extends on the functionality in decomposeHomographyMat, making it possible to reduce the solution set from 4 to 2 possible solutions if additional information is present.

### CODE REVIEW

File: modules/calib3d/include/opencv2/calib3d.hpp [hunk 1]
```diff
@@ -1598,6 +1598,31 @@ CV_EXPORTS_W int decomposeHomographyMat(InputArray H,
...
+homographies are consistent with all visible reference points being in front of the camera. The inputs
+are left unchanged; the filtered solution set is returned as indices into the existing one.
+
+*/
+CV_EXPORTS_W std::vector<int> filterHomographyDecompSolutionsByPointNormals(
+    const std::vector<Mat>& rotations,
+    const std::vector<Mat>& normals,
+    const std::vector<Point2f>& beforeRectifiedPoints,
+    const std::vector<Point2f>& afterRectifiedPoints,
+    const Mat& mask);
```
Comment (sovrasov) [inline modules/calib3d/include/opencv2/calib3d.hpp]: cv::InputArray and cv::OutputArray should be used to pass or return parameters

**Comment by sovrasov** [comment]
@Vaesper why new files are needed for this function?
It's better to keep declaration in calib3d.hpp and move implementation to homography_decomp.cpp

**Comment by sovrasov** [comment]
Also will be useful to add simple test for this function.

**Comment by Vaesper** [comment]
@sovrasov new files are not /needed/, but I could not easily make sense of the way homography_decomp.cpp was organized. It has, for example, no corresponding .h file and so instead of messing up the way that file was organized and making it unrecognizable to the person/people who worked on it, I decided to put my stuff in a different file. Cleaner and simpler, imho.

What kind of test would you like to see? I have some code of it working in practice.

I will work on making the function parameters in line with the InputArray/OutputArray datatypes.

**Comment by vpisarev** [comment]
@Vaesper, thanks for the contribution!

On the tests: the code put to OpenCV stays there for years; for each piece of new functionality we need some sort of regression test to check that the code still works. We cannot ask you to check this function after each modification that we make in OpenCV. So the test is obligatory.

@sovrasov's comment on putting this code to homography_decomp.cpp is quite valid; these functions are very similar in purpose. At the same time, I do not think we need extra internal header, because the function is exposed in the external header. If the external function API is not convenient enough, we should think of changing the external API rather than declaring the internal function in internal header.

I also suggest to shorten the function name as "filterHomographyDecompByNormals". Also, the output vector<int> should be made the output parameter instead of return value

**Comment by Vaesper** [comment]
@vpisarev 

I see, that makes sense, but it doesn't make it clear to me exactly what kind of test must be written and in what format. Is/are there some other test(s) that I can look at for inspiration?

As for the header file, I see your point, and indeed opencv compiles perfectly on my machine with my .h file removed. One thing that prevents me from merging my code with homography_decomp.cpp is the attribution to Samson Yilma just above and in the license agreement. I'm not sure what to do with that, as I don't want to publish code under that name but also do not wish to claim credit for their contributions by adding my name or something.

**Comment by Vaesper** [comment]
As for the name of the function, I think a more intuitively clear one would be filterHomographyDecompByVisibleRefpoints as what it does is verify that the reference points (the ones used to estimate the homography in the first place) are actually in front of the camera according to the solution. If that is too long, perhaps filterHomographyDecompByRefpoints as those are the extra parameters required.

File: modules/calib3d/include/opencv2/calib3d.hpp [hunk 2]
```diff
@@ -1598,6 +1598,30 @@ CV_EXPORTS_W int decomposeHomographyMat(InputArray H,
...
+This function is intended to filter the output of the decomposeHomographyMat based on additional
+information as described in @cite Malis . The summary of the method: the decomposeHomographyMat function
+returns 2 unique solutions and their "opposites" for a total of 4 solutions. If we have access to the
+sets of points visible in the camera frame before and after the homography transformation is applied,
+we can determine which are the true potential solutions and which are the opposites by verifying which
+homographies are consistent with all visible reference points being in front of the camera. The inputs
+are left unchanged; the filtered solution set is returned as indices into the existing one.
+
+*/
+CV_EXPORTS_W Mat filterHomographyDecompByVisibleRefpoints(InputArrayOfArrays rotations,
```
Comment (sovrasov) [inline modules/calib3d/include/opencv2/calib3d.hpp]: Please change function prototype to
`
CV_EXPORTS_W void filterHomographyDecompByVisibleRefpoints(InputArrayOfArrays rotations,
                                                          InputArrayOfArrays normals,
                                                          InputArray beforePoints,
                                                          InputArray afterPoints,
                                                         OutputArray possibleSolutions,
                                                          InputArray pointsMask = noArray()
                                                          );
`
It needed for correct generation of python bindings.

**Comment by sovrasov** [comment]
The simplest way to write a test is to prepare some synthetic input data and compare function output with expectations.
Possible test pipeline:
predefined H and K -> decomposeHomographyMat -> decomposition results, normals -> filterHomographyDecompByVisibleRefpoints -> check output

**Comment by sovrasov** [comment]
@Vaesper what about test and moving implementaion to  homography_decomp.cpp?

**Comment by Vaesper** [comment]
In progress.

**Comment by Vaesper** [comment]
I am having trouble implementing the test; I have been setting a breakpoint in my prototype program and copying over values into the test, but for some reason I am not getting the same output from the function as I do in the prototype program. I have been looking into various possible causes but have been unable to figure out what is going wrong.

**Comment by sovrasov** [comment]
@Vaesper commit code with errors, I will look at it.

**Comment by Vaesper** [comment]
The data hardcoded into the test database is lifted from a prototype program running a BRISK detector (`BRISK::create(50, 4, 1.0f);`) and an ORB feature extractor (`ORB::create();`), using a `BFMatcher` with `NORM_HAMMING`.

I can provide the two images used to get these results if you can point me to a location where I can upload them without worrying about them being recompressed (not sure if that would be a relevant consideration but I'd rather be certain)..

**Comment by sovrasov** [comment]
I think that it's better to generate of some synthetic data for test.
But let's try to fix your version. You can post pictures right here (github won't recompress them).

**Comment by sovrasov** [comment]
BTW current version of test has been passed on Win10x64 builder. Is it wrong result?

**Comment by Vaesper** [comment]
That's... Intriguing. I notice the other failing builds seem to be mostly about "truncation from 'double' to 'float'" warnings. Perhaps I can try and resolve those warnings and see if the tests pass successfully in the other systems if I do so. Maybe (hopefully) it's actually my system somehow not running the test correctly.

**Comment by Vaesper** [comment]
Seems like there was some C++11 vs C++98 issue; after turning the direct vector initializations into array initializations with subsequent conversion to vectors, a lot of the issues went away. Apparently this issue did not cause the build to fail on my system, but did cause incorrect behaviour leading to the failed test.

There's still compile errors for some platforms, with a cursory glance over the logs I was not able to determine the problem. I will look into it when I have the time.

**Comment by Vaesper** [comment]
Is there a way to request a rebuild for certain systems? Logs have been cleared but I don't need to use up resources to rebuild the successful builds.

**Comment by sovrasov** [comment]
@Vaesper there is no any special ways for it. I restarted your build.

**Comment by Vaesper** [comment]
Builds/tests are starting to look good, only iOS one remaining. Should I try to flatten history once things are in passing state?

**Comment by sovrasov** [comment]
Yes, also you should move implementation and test into sufficient files and also move hardcoded vectors into github.com/opencv/opencv_extra/ to consist with existing coding style.

**Comment by vpisarev** [comment]
superseded by #11846

**Pull request closed without merging**